File: crypto/src/main/java/io/oversec/one/crypto/ui/NewPasswordInputDialog.java
package io.oversec.one.crypto.ui;

import android.content.*;
import android.graphics.PorterDuff;
import android.graphics.PorterDuffColorFilter;
import android.os.IBinder;
import android.os.RemoteException;
import android.support.annotation.NonNull;
import android.support.design.widget.TextInputLayout;
import android.support.v4.content.ContextCompat;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.View;
import android.view.WindowManager;
import android.view.inputmethod.EditorInfo;
import android.widget.*;
import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;
import io.oversec.one.crypto.IZxcvbnService;
import io.oversec.one.crypto.R;
import io.oversec.one.crypto.ZxcvbnResult;
import io.oversec.one.crypto.ui.util.EditTextPasswordWithVisibilityToggle;
import uk.co.biddell.diceware.dictionaries.DiceWare;
import uk.co.biddell.diceware.dictionaries.DiceWords;

import java.io.IOException;

public class NewPasswordInputDialog {
    private static final int ENTROPY_MEDIUM = 45;
    private static final int ENTROPY_HIGH_SHARE = 75;
    private static final int ENTROPY_HIGH_PBKDF = 75;
    private static final int ENTROPY_HIGH_DEVICE = 60;


    private static final int DICEWARE_WORDS_KEYSTORE = 4;
    private static final int DICEWARE_WORDS_SHARE = 5;
    private static final int DICEWARE_WORDS_PBKDF = 6;

    public enum MODE {
        SHARE, KEYSTORE, PBKDF
    }

    public static void show(final Context ctx, final MODE mode, final NewPasswordInputDialogCallback callback) {
        Intent serviceIntent = new Intent()
                .setComponent(new ComponentName(ctx, "io.oversec.one.crypto.ZxcvbnService")); //do NOT reference by CLASS
        ctx.startService(serviceIntent);
        final IZxcvbnService[] mZxcvbnService = new IZxcvbnService[1];
        final ServiceConnection mConnection = new ServiceConnection() {
            @Override
            public void onServiceConnected(ComponentName name, IBinder service) {
                mZxcvbnService[0] = IZxcvbnService.Stub.asInterface(service);
            }

            @Override
            public void onServiceDisconnected(ComponentName name) {
                mZxcvbnService[0] = null;
            }
        };
        ctx.bindService(serviceIntent, mConnection, Context.BIND_ALLOW_OOM_MANAGEMENT);

        final MaterialDialog dialog = new MaterialDialog.Builder(ctx)

                .customView(R.layout.new_password_input_dialog, false)
                .positiveText(getPositiveText(mode))
                .neutralText(R.string.common_cancel)
                .autoDismiss(false)
                .dismissListener(new DialogInterface.OnDismissListener() {
                    @Override
                    public void onDismiss(DialogInterface dialog) {
                        try {
                            if (mZxcvbnService[0]!=null) {
                                mZxcvbnService[0].exit();
                            }
                        } catch (RemoteException e) {
                            e.printStackTrace();
                        }
                        ctx.unbindService(mConnection);
                    }
                })
                .cancelListener(new DialogInterface.OnCancelListener() {
                    @Override
                    public void onCancel(DialogInterface dialog) {
                        callback.neutralAction();
                        //TODO: clear passwords on cancel?
                        dialog.dismiss();
                    }
                })
                .onPositive(new MaterialDialog.SingleButtonCallback() {
                    @Override
                    public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                        View view = dialog.getCustomView();
                        if (handlePositive(view, callback, mode, mZxcvbnService[0])) {
                            dialog.dismiss();
                        }
                    }

                })
                .onNeutral(new MaterialDialog.SingleButtonCallback() {
                    @SuppressWarnings("ConstantConditions")
                    @Override
                    public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                        View view = dialog.getCustomView();
                        EditText etPw1 = (EditText) view.findViewById(R.id.new_password_password);
                        EditText etPw2 = (EditText) view.findViewById(R.id.new_password_password_again);
                        etPw1.setText(""); //TODO better way?
                        etPw2.setText(""); //TODO better way?
                        callback.neutralAction();
                        dialog.dismiss();
                    }
                }).build();

        dialog.getWindow().setSoftInputMode(
                WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);

        dialog.getWindow().setLayout(WindowManager.LayoutParams.MATCH_PARENT,
                WindowManager.LayoutParams.WRAP_CONTENT);

        final View view = dialog.getCustomView();

        final EditTextPasswordWithVisibilityToggle etPw1 = (EditTextPasswordWithVisibilityToggle) view.findViewById(R.id.new_password_password);
        final EditTextPasswordWithVisibilityToggle etPw2 = (EditTextPasswordWithVisibilityToggle) view.findViewById(R.id.new_password_password_again);

        etPw2.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                if (actionId == EditorInfo.IME_ACTION_DONE) {
                    if (handlePositive(view, callback, mode, mZxcvbnService[0])) {
                        dialog.dismiss();
                        return true;
                    }
                }
                return false;
            }
        });

        final TextInputLayout wrapPw1 = (TextInputLayout) view.findViewById(R.id.new_password_password_wrapper);

        TextView tvText = (TextView) view.findViewById(R.id.new_password_text);
        tvText.setText(getBody(mode));

        TextView tvTitle = (TextView) view.findViewById(R.id.new_password_title);
        tvTitle.setText(getTitle(mode));

        CheckBox cbWeak = (CheckBox) view.findViewById(R.id.cb_accept_weak_password);
        cbWeak.setVisibility(View.GONE);

        Button btSuggest = (Button) view.findViewById(R.id.new_password_generate);
        btSuggest.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //  cbShowPassphrase.setChecked(true);
                try {
                    DiceWords dw = new DiceWare(ctx).getDiceWords(
                            getDicewareExtraSecurity(mode),
                            getDicewareNumWords(mode));

                    etPw1.setText(dw.toString());
                    etPw1.setPasswordVisible(true);
                    etPw2.setPasswordVisible(true);
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }


        });

        final SeekBar sbStrength = (SeekBar) view.findViewById(R.id.create_key_seekbar);
        etPw1.addTextChangedListener(new TextWatcher() {


            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {
                int entropy = calcPasswordEntropy(s, wrapPw1, mZxcvbnService[0]);
                updateSeekBar(view, sbStrength, entropy, mode);
            }

            @Override
            public void afterTextChanged(Editable s) {

            }
        });

        updateSeekBar(view, sbStrength, 0, mode);
        dialog.show();
    }


    private static int getBody(MODE mode) {
        switch (mode) {
            case KEYSTORE:
                return R.string.new_password_keystore_text;
            case PBKDF:
                return R.string.new_password_pbkdf_text;
            case SHARE:
                return R.string.new_password_share_text;
        }
        return 0;
    }

    private static int getTitle(MODE mode) {
        switch (mode) {
            case KEYSTORE:
                return R.string.new_password_keystore_title;
            case PBKDF:
                return R.string.new_password_pbkdf_title;
            case SHARE:
                return R.string.new_password_share_title;
        }
        return 0;
    }

    private static DiceWare.Type getDicewareExtraSecurity(MODE mode) {
        switch (mode) {
            case KEYSTORE:
                return DiceWare.Type.PASSPHRASE;
            case PBKDF:
                return DiceWare.Type.PASSPHRASE_EXTRA_SECURITY;
            case SHARE:
                return DiceWare.Type.PASSPHRASE_EXTRA_SECURITY;
        }
        return null;
    }

    private static int getDicewareNumWords(MODE mode) {
        switch (mode) {
            case KEYSTORE:
                return DICEWARE_WORDS_KEYSTORE;
            case PBKDF:
                return DICEWARE_WORDS_PBKDF;
            case SHARE:
                return DICEWARE_WORDS_SHARE;
        }
        return 0;
    }

    private static int getPositiveText(MODE mode) {
        switch (mode) {
            case KEYSTORE:
                return R.string.action_save;
            case PBKDF:
                return R.string.action_generate;
            case SHARE:
                return R.string.action_share;
        }
        return 0;

    }


    private static int getEntropyHighLevel(MODE mode) {
        switch (mode) {
            case KEYSTORE:
                return ENTROPY_HIGH_DEVICE;
            case PBKDF:
                return ENTROPY_HIGH_PBKDF;
            case SHARE:
                return ENTROPY_HIGH_SHARE;
        }
        return 0;
    }

    private static int getEntropyMinimum(MODE mode) {
        switch (mode) {
            case KEYSTORE:
                return ENTROPY_MEDIUM; //facilitate usage fur "dumb" users
            case PBKDF:
                return ENTROPY_MEDIUM; //facilitate usage fur "dumb" users
            case SHARE:
                return ENTROPY_HIGH_SHARE;
        }
        return 0;
    }

    private static void updateSeekBar(View view, SeekBar sbStrength, int entropy, MODE mode) {

        sbStrength.setMax(100);
        sbStrength.setProgress(Math.max(10, entropy));
        int color = R.color.password_strength_low;
        if (entropy >= ENTROPY_MEDIUM) {
            color = R.color.password_strength_medium;
        }


        if (entropy >= getEntropyHighLevel(mode)) {
            color = R.color.password_strength_high;
            CheckBox cbWeak = (CheckBox) view.findViewById(R.id.cb_accept_weak_password);
            cbWeak.setVisibility(View.GONE);
        }
        sbStrength.getProgressDrawable().setColorFilter(new PorterDuffColorFilter(ContextCompat.getColor(sbStrength.getContext(), color), PorterDuff.Mode.MULTIPLY));


    }


    private static int calcPasswordEntropy(CharSequence s, TextInputLayout wrapper, IZxcvbnService zxcvbn) {
        if (zxcvbn==null) {
            //service not bound?
            return 0;
        }
        try {
            ZxcvbnResult r = zxcvbn.calcEntropy(s.toString());

            wrapper.setError(r.getWarning());

            int res = r.getEntropy();
            return res;
        } catch (RemoteException ex) {
            ex.printStackTrace();
            return 0;
        }
    }



    private static boolean handlePositive(View view, NewPasswordInputDialogCallback callback, MODE mode, IZxcvbnService zxcvbn) {
        EditText etPw1 = (EditText) view.findViewById(R.id.new_password_password);
        EditText etPw2 = (EditText) view.findViewById(R.id.new_password_password_again);
        TextInputLayout wrapPw1 = (TextInputLayout) view.findViewById(R.id.new_password_password_wrapper);
        TextInputLayout wrapPw2 = (TextInputLayout) view.findViewById(R.id.new_password_password_again_wrapper);
        CheckBox cbWeak = (CheckBox) view.findViewById(R.id.cb_accept_weak_password);
        Editable editablePw1 = etPw1.getText();
        Editable editablePw2 = etPw2.getText();

        if (!editablePw1.toString().equals(editablePw2.toString())) {
            wrapPw1.setError(view.getContext().getString(R.string.error_passwords_dont_match));
            return false;
        }


        int entropy = calcPasswordEntropy(etPw1.getText(), wrapPw1, zxcvbn);

        if (entropy < getEntropyMinimum(mode) && !cbWeak.isChecked()) {
            wrapPw1.setError(view.getContext().getString(R.string.error_password_length));

            cbWeak.setVisibility(View.VISIBLE);
            cbWeak.requestFocus();
            cbWeak.getParent().requestChildFocus(cbWeak, cbWeak);
            return false;
        }


        int pl = editablePw1.length();
        char[] aPassPhrase = new char[pl];
        editablePw1.getChars(0, pl, aPassPhrase, 0);

        etPw1.setText(""); //TODO better way?
        etPw2.setText(""); //TODO better way?

        callback.positiveAction(aPassPhrase);


        return true;
    }


}

File: crypto/src/main/java/io/oversec/one/crypto/ui/NewPasswordInputDialog.kt
package io.oversec.one.crypto.ui

import android.content.*
import android.graphics.PorterDuff
import android.graphics.PorterDuffColorFilter
import android.os.IBinder
import android.os.RemoteException
import android.support.design.widget.TextInputLayout
import android.support.v4.content.ContextCompat
import android.text.Editable
import android.text.TextWatcher
import android.view.View
import android.view.WindowManager
import android.view.inputmethod.EditorInfo
import android.widget.*
import com.afollestad.materialdialogs.MaterialDialog
import io.oversec.one.crypto.IZxcvbnService
import io.oversec.one.crypto.R
import io.oversec.one.crypto.ui.util.EditTextPasswordWithVisibilityToggle
import uk.co.biddell.diceware.dictionaries.DiceWare
import java.io.IOException

object NewPasswordInputDialog {
    private const val ENTROPY_MEDIUM = 45
    private const val ENTROPY_HIGH_SHARE = 75
    private const val ENTROPY_HIGH_PBKDF = 75
    private const val ENTROPY_HIGH_DEVICE = 60

    private const val DICEWARE_WORDS_KEYSTORE = 4
    private const val DICEWARE_WORDS_SHARE = 5
    private const val DICEWARE_WORDS_PBKDF = 6

    enum class MODE {
        SHARE, KEYSTORE, PBKDF
    }

    fun show(ctx: Context, mode: MODE, callback: NewPasswordInputDialogCallback) {
        val serviceIntent = Intent()
            .setComponent(
                ComponentName(
                    ctx,
                    "io.oversec.one.crypto.ZxcvbnService"
                )
            ) //do NOT reference by CLASS
        ctx.startService(serviceIntent)
        val mZxcvbnService = arrayOfNulls<IZxcvbnService>(1)
        val mConnection = object : ServiceConnection {
            override fun onServiceConnected(name: ComponentName, service: IBinder) {
                mZxcvbnService[0] = IZxcvbnService.Stub.asInterface(service)
            }

            override fun onServiceDisconnected(name: ComponentName) {
                mZxcvbnService[0] = null
            }
        }
        ctx.bindService(serviceIntent, mConnection, Context.BIND_ALLOW_OOM_MANAGEMENT)

        val dialog = MaterialDialog.Builder(ctx)
            .customView(R.layout.new_password_input_dialog, false)
            .positiveText(getPositiveText(mode))
            .neutralText(R.string.common_cancel)
            .autoDismiss(false)
            .dismissListener {
                try {
                    mZxcvbnService[0]?.exit()
                } catch (e: RemoteException) {
                    e.printStackTrace()
                }

                ctx.unbindService(mConnection)
            }
            .cancelListener { dialog ->
                callback.neutralAction()
                //TODO: clear passwords on cancel?
                dialog.dismiss()
            }
            .onPositive { dialog, which ->
                val view = dialog.customView
                if (handlePositive(view!!, callback, mode, mZxcvbnService[0])) {
                    dialog.dismiss()
                }
            }
            .onNeutral { dialog, which ->
                val view = dialog.customView
                val etPw1 = view!!.findViewById<View>(R.id.new_password_password) as EditText
                val etPw2 = view.findViewById<View>(R.id.new_password_password_again) as EditText
                etPw1.setText("") //TODO better way?
                etPw2.setText("") //TODO better way?
                callback.neutralAction()
                dialog.dismiss()
            }.build()

        dialog.window!!.setSoftInputMode(
            WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE
        )

        dialog.window!!.setLayout(
            WindowManager.LayoutParams.MATCH_PARENT,
            WindowManager.LayoutParams.WRAP_CONTENT
        )

        val view = dialog.customView

        val etPw1 =
            view!!.findViewById<View>(R.id.new_password_password) as EditTextPasswordWithVisibilityToggle
        val etPw2 =
            view.findViewById<View>(R.id.new_password_password_again) as EditTextPasswordWithVisibilityToggle

        etPw2.setOnEditorActionListener(TextView.OnEditorActionListener { v, actionId, event ->
            if (actionId == EditorInfo.IME_ACTION_DONE) {
                if (handlePositive(view, callback, mode, mZxcvbnService[0])) {
                    dialog.dismiss()
                    return@OnEditorActionListener true
                }
            }
            false
        })

        val wrapPw1 = view.findViewById<View>(R.id.new_password_password_wrapper) as TextInputLayout

        val tvText = view.findViewById<View>(R.id.new_password_text) as TextView
        tvText.setText(getBody(mode))

        val tvTitle = view.findViewById<View>(R.id.new_password_title) as TextView
        tvTitle.setText(getTitle(mode))

        val cbWeak = view.findViewById<View>(R.id.cb_accept_weak_password) as CheckBox
        cbWeak.visibility = View.GONE

        val btSuggest = view.findViewById<View>(R.id.new_password_generate) as Button
        btSuggest.setOnClickListener {
            //  cbShowPassphrase.setChecked(true);
            try {
                val dw = DiceWare(ctx).getDiceWords(
                    getDicewareExtraSecurity(mode)!!,
                    getDicewareNumWords(mode)
                )

                etPw1.setText(dw.toString())
                etPw1.setPasswordVisible(true)
                etPw2.setPasswordVisible(true)
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }

        val sbStrength = view.findViewById<View>(R.id.create_key_seekbar) as SeekBar
        etPw1.addTextChangedListener(object : TextWatcher {

            override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {
            }

            override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
                val entropy = calcPasswordEntropy(s, wrapPw1, mZxcvbnService[0])
                updateSeekBar(view, sbStrength, entropy, mode)
            }

            override fun afterTextChanged(s: Editable) {
            }
        })

        updateSeekBar(view, sbStrength, 0, mode)
        dialog.show()
    }


    private fun getBody(mode: MODE): Int {
        return when (mode) {
            NewPasswordInputDialog.MODE.KEYSTORE -> R.string.new_password_keystore_text
            NewPasswordInputDialog.MODE.PBKDF -> R.string.new_password_pbkdf_text
            NewPasswordInputDialog.MODE.SHARE -> R.string.new_password_share_text
        }
    }

    private fun getTitle(mode: MODE): Int {
        return when (mode) {
            NewPasswordInputDialog.MODE.KEYSTORE -> R.string.new_password_keystore_title
            NewPasswordInputDialog.MODE.PBKDF -> R.string.new_password_pbkdf_title
            NewPasswordInputDialog.MODE.SHARE -> R.string.new_password_share_title
        }
    }

    private fun getDicewareExtraSecurity(mode: MODE): DiceWare.Type? {
        return when (mode) {
            NewPasswordInputDialog.MODE.KEYSTORE -> DiceWare.Type.PASSPHRASE
            NewPasswordInputDialog.MODE.PBKDF -> DiceWare.Type.PASSPHRASE_EXTRA_SECURITY
            NewPasswordInputDialog.MODE.SHARE -> DiceWare.Type.PASSPHRASE_EXTRA_SECURITY
        }
    }

    private fun getDicewareNumWords(mode: MODE): Int {
        return when (mode) {
            NewPasswordInputDialog.MODE.KEYSTORE -> DICEWARE_WORDS_KEYSTORE
            NewPasswordInputDialog.MODE.PBKDF -> DICEWARE_WORDS_PBKDF
            NewPasswordInputDialog.MODE.SHARE -> DICEWARE_WORDS_SHARE
        }
    }

    private fun getPositiveText(mode: MODE): Int {
        return when (mode) {
            NewPasswordInputDialog.MODE.KEYSTORE -> R.string.action_save
            NewPasswordInputDialog.MODE.PBKDF -> R.string.action_generate
            NewPasswordInputDialog.MODE.SHARE -> R.string.action_share
        }
    }


    private fun getEntropyHighLevel(mode: MODE): Int {
        return when (mode) {
            NewPasswordInputDialog.MODE.KEYSTORE -> ENTROPY_HIGH_DEVICE
            NewPasswordInputDialog.MODE.PBKDF -> ENTROPY_HIGH_PBKDF
            NewPasswordInputDialog.MODE.SHARE -> ENTROPY_HIGH_SHARE
        }
    }

    private fun getEntropyMinimum(mode: MODE): Int {
        return when (mode) {
            NewPasswordInputDialog.MODE.KEYSTORE -> ENTROPY_MEDIUM //facilitate usage fur "dumb" users
            NewPasswordInputDialog.MODE.PBKDF -> ENTROPY_MEDIUM //facilitate usage fur "dumb" users
            NewPasswordInputDialog.MODE.SHARE -> ENTROPY_HIGH_SHARE
        }
    }

    private fun updateSeekBar(view: View?, sbStrength: SeekBar, entropy: Int, mode: MODE) {

        sbStrength.max = 100
        sbStrength.progress = Math.max(10, entropy)
        var color = R.color.password_strength_low
        if (entropy >= ENTROPY_MEDIUM) {
            color = R.color.password_strength_medium
        }


        if (entropy >= getEntropyHighLevel(mode)) {
            color = R.color.password_strength_high
            val cbWeak = view!!.findViewById<View>(R.id.cb_accept_weak_password) as CheckBox
            cbWeak.visibility = View.GONE
        }
        sbStrength.progressDrawable.colorFilter = PorterDuffColorFilter(
            ContextCompat.getColor(sbStrength.context, color),
            PorterDuff.Mode.MULTIPLY
        )


    }


    private fun calcPasswordEntropy(
        s: CharSequence,
        wrapper: TextInputLayout,
        zxcvbn: IZxcvbnService?
    ): Int {
        zxcvbn?: return 0  //service not bound?
        return try {
            val r = zxcvbn.calcEntropy(s.toString())
            wrapper.error = r.warning
            r.entropy
        } catch (ex: RemoteException) {
            ex.printStackTrace()
            0
        }

    }


    private fun handlePositive(
        view: View,
        callback: NewPasswordInputDialogCallback,
        mode: MODE,
        zxcvbn: IZxcvbnService?
    ): Boolean {
        val etPw1 = view.findViewById<View>(R.id.new_password_password) as EditText
        val etPw2 = view.findViewById<View>(R.id.new_password_password_again) as EditText
        val wrapPw1 = view.findViewById<View>(R.id.new_password_password_wrapper) as TextInputLayout
        val wrapPw2 =
            view.findViewById<View>(R.id.new_password_password_again_wrapper) as TextInputLayout
        val cbWeak = view.findViewById<View>(R.id.cb_accept_weak_password) as CheckBox
        val editablePw1 = etPw1.text
        val editablePw2 = etPw2.text

        if (editablePw1.toString() != editablePw2.toString()) {
            wrapPw1.error = view.context.getString(R.string.error_passwords_dont_match)
            return false
        }


        val entropy = calcPasswordEntropy(etPw1.text, wrapPw1, zxcvbn)

        if (entropy < getEntropyMinimum(mode) && !cbWeak.isChecked) {
            wrapPw1.error = view.context.getString(R.string.error_password_length)

            cbWeak.visibility = View.VISIBLE
            cbWeak.requestFocus()
            cbWeak.parent.requestChildFocus(cbWeak, cbWeak)
            return false
        }


        val pl = editablePw1.length
        val aPassPhrase = CharArray(pl)
        editablePw1.getChars(0, pl, aPassPhrase, 0)

        etPw1.setText("") //TODO better way?
        etPw2.setText("") //TODO better way?

        callback.positiveAction(aPassPhrase)


        return true
    }


}
